File: CodeWars/Fundamentals/CompleteThePattern/CompleteThePattern12.cpp
#include <iostream>
#include <string>
#include <cassert>

std::string pattern (int n) {
    if (n < 1)
        return "";

    std::string res = "";
    std::string temp (2*n-1, ' ');

    for (int i = 0 ; i < n ; ++i) {
        temp[i] = '0'+((i+1)%10);
        temp[2*n-2-i] = '0'+((i+1)%10);
        res += temp;
        res += '\n';
        temp = std::string (2*n-1, ' ');
    }
    for (int i = n-2 ; i >= 0 ; --i) {
        temp[i] = '0'+((i+1)%10);
        temp[2*n-2-i] = '0'+((i+1)%10);
        res += temp;
        res += '\n';
        temp = std::string (2*n-1, ' ');
    }
    res.pop_back ();

    return res;
}

void tests () {
    std::string expected = "\n"
        "1   1\n"
        " 2 2 \n"
        "  3  \n"
        " 2 2 \n"
        "1   1";
    std::string actual = "\n" + pattern(3);
    assert (expected == actual);

    expected = "\n"
        "1       1\n"
        " 2     2 \n"
        "  3   3  \n"
        "   4 4   \n"
        "    5    \n"
        "   4 4   \n"
        "  3   3  \n"
        " 2     2 \n"
        "1       1";
    actual = "\n" + pattern(5);
    assert (expected == actual);
}

int main () {
    tests ();

    return 0;
}

File: CodeWars/Fundamentals/CompleteThePattern/CompleteThePattern12.py
import unittest


def pattern(n):
    if n < 1:
        return ""
    res = ""
    temp = list(' '*(2*n-1))

    for i in range(n):
        temp[i] = str((i+1) % 10)
        temp[2*n-2-i] = str((i+1) % 10)
        res += "".join(temp) + '\n'
        temp = list(' '*(2*n-1))

    for i in range(n-2, -1, -1):
        temp[i] = str((i+1) % 10)
        temp[2*n-2-i] = str((i+1) % 10)
        res += "".join(temp) + '\n'
        temp = list(' '*(2*n-1))

    return res[:-1]


class CompleteThePatternTest(unittest.TestCase):
    def testCompletePattern(self):
        self.assertEquals(pattern(3), "1   1\n 2 2 \n  3  \n 2 2 \n1   1")
        self.assertEquals(pattern(15), "1                           1\n 2                         2 \n  3                       3  \n   4                     4   \n    5                   5    \n     6                 6     \n      7               7      \n       8             8       \n        9           9        \n         0         0         \n          1       1          \n           2     2           \n            3   3            \n             4 4             \n              5              \n             4 4             \n            3   3            \n           2     2           \n          1       1          \n         0         0         \n        9           9        \n       8             8       \n      7               7      \n     6                 6     \n    5                   5    \n   4                     4   \n  3                       3  \n 2                         2 \n1                           1")


suite = unittest.TestLoader().loadTestsFromTestCase(CompleteThePatternTest)
unittest.TextTestRunner(verbosity=2).run(suite)

print(pattern(3))
print(pattern(5))
